proposals/custom-equals-in-value-classes.md

```diff
@@ -0,0 +1,262 @@
+
+> From now, we will be calling equals from `Any` *untyped* equals, in contrast, to *typed* one.
+
+### Interaction of typed and untyped equals
+
+We need to establish consistency of typed and untyped equals, i.e. the following property:
+> For any `x`, `y` such that `x`, `y` refer instances of an inline class, it must be true
+> that `x.equals(y) == (x as Any).equals(y)`
+
+We propose the following:
+
+* If only typed equals is declared:
+    * Generate consistent untyped one. Implementation of untyped equals will be trying to cast the argument to
+      corresponding value class type and passing to typed equals if succeeded.
+* If only untyped equals is declared:
+    * Generate consistent typed one. Implementation of typed equals will be boxing argument and passing it to untyped
+      equals. Implement special diagnostics to warn programmers about negative performance impact of boxing in
+      auto-generated code
+* If both equals methods are declared:
+    * Consider programmer responsible for their consistency
+
+Note, that in the first case on JVM level we will be casting argument to the raw type of value class, thus we will know
+nothing about its type arguments. We emphasize this fact by requiring the parameter of typed equals to be
+star-projection of the value class.
+
+### Custom `hashCode`
+
+It is obvious that the ability of overriding `equals` makes us allow overriding of `hashCode`.
+
+### `@TypedEquals` and `@AllowTypedEquals` annotations
+
+Customization of equals in value classes is currently considered as an experimental feature. Thus, we want to force
+user to `@OptIn` before using it. In previous section we introduced `@TypedEquals` annotation, which serves as a marker
+for typed equals function. Unfortunately, simply annotating `@TypedEquals` with `@RequireOptIn` will not work, as
+typed equals may be invoked implicitly:
+
+```Kotlin
+@JvmInline
+value class IC(val x: Int) {
+    @TypedEquals
+    fun equals(other: IC) = ...
+}
+
+fun foo() {
+    setOf(IC(1), IC(2)) // compiles without @OptIn, but invokes typed equals internally
+}
+```
+
+We propose to add new annotation `@AllowTypedEquals`. This annotation is applicable for value classes only and any value
+class that declares typed equal must be annotated by `@AllowTypedEquals`:
+
+```Kotlin
+@Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@RequiresOptIn
+annotation class AllowTypedEquals
+```
+
+```Kotlin
+@JvmInline
+@AllowTypedEquals
+value class IC(val x: Int) {
+    @TypedEquals
+    fun equals(other: IC) = ...
+}
+
+@OptIn(AllowTypedEquals::class) // usage of value class IC requires OptIn
+fun foo() {
+    setOf(IC(1), IC(2))
+}
+```
+
+## Other questions
+
+### Inline classes over inline classes
+
+Suppose we have an inline class `B` over inline class `A` and `A` declares typed equals. In this case,
+default-generated equals in `B` will be comparing it underlying values using custom equals of `A`.
```

Comment:
Will `B` require opt-in?

Reply:
Sure. Any usage of `A` will require opt-in and `B` will use `A` for declaring its primary constructor.